src/Connection/Http1Connection.php

```diff
@@ -332,26 +368,44 @@ private function normalizeTraceRequest(Request $request): Request
         return $request;
     }
 
-    private function shouldCloseSocketAfterResponse(Response $response): bool
+    private function determineKeepAliveTimeout(Response $response): int
     {
         $request = $response->getRequest();
 
         $requestConnHeader = $request->getHeader('connection');
         $responseConnHeader = $response->getHeader('connection');
 
-        if ($requestConnHeader && !\strcasecmp($requestConnHeader, 'close')) {
-            return true;
+        if (!\strcasecmp($requestConnHeader, 'close')) {
+            return 0;
+        }
+
+        if ($response->getProtocolVersion() === '1.0') {
+            return 0;
+        }
+
+        if (\strcasecmp($responseConnHeader, 'keep-alive')) {
+            return 0;
         }
 
-        if ($responseConnHeader && !\strcasecmp($responseConnHeader, 'close')) {
-            return true;
+        $responseKeepAliveHeader = $response->getHeader('keep-alive');
+
+        if ($responseKeepAliveHeader === null) {
+            return self::MAX_KEEP_ALIVE_TIMEOUT;
+        }
+
+        $parts = \array_map('trim', \explode(',', $responseKeepAliveHeader));
+
+        $params = [];
+        foreach ($parts as $part) {
+            [$key, $value] = \array_map('trim', \explode('=', $part)) + [null, null];
+            $params[$key] = (int) $value;
         }
 
-        if (!$responseConnHeader && $response->getProtocolVersion() === '1.0') {
-            return true;
+        if ($this->requestCounter >= $params['max'] ?? \PHP_INT_MAX) {
+            return 0;
         }
 
-        return false;
+        return \min(\max(0, $params['timeout'] ?? 0), self::MAX_KEEP_ALIVE_TIMEOUT);
```

Comment:
Why is `0` the default value here?

Reply:
It's only `0` if the timeout value is negative or unspecified. I'm not sure if that's correct, can a `keep-alive` header have only a `max` value but no `timeout` value?

Reply:
`max` can be ignored if no pipelining is used: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Keep-Alive

Reply:
Cool, that simplifies things a bit.